Refactor of batch job kickoff and tracking process

There is actually an ERROR message just a little further down one of the Ultra logs.
```
[ERROR] KeyError: 'job_definition'
Traceback (most recent call last):
  File "/var/task/SDSCode/batch_starter.py", line 399, in lambda_handler
    update_status_table(status_params)
  File "/var/task/SDSCode/database_handler.py", line 60, in update_status_table
    result.job_definition = status_params["job_definition"]
```
This is coming from within update_status_table
[https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/d6f2d7c67e[…]108b2a/sds_data_manager/lambda_code/SDSCode/database_handler.py](https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/d6f2d7c67ed7f1acb3d11b9f7ff77030ba108b2a/sds_data_manager/lambda_code/SDSCode/database_handler.py#L47-L64)
and we are actually hitting the first path and writing the record in the first log. Then going down the other path in the second log. But then, we are getting a Python exception thrown here because the job_definition key doesn't exist, and that exception kills the lambda and therefore we don't actually get the second batch job somewhat luckily :slightly_smiling_face: So that is why there wasn't two batch jobs kicked off, even though theoretically we should have.
All of this being said, I think we should do multiple changes to address this:
* Reduce the number of DB connections we are making and pass the session around to our other functions more often. So request a DB connection right away in the lambda handler and then pass it through. (Right now we are making a session in batch_starter and a second one in update_status_table, and that added a pretty significant delay of ~0.5s)
* Update our Unique Constraint on the database to include all of the specific properties that we expect to be unique for each job so we can't get duplicates (remove ID). [https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/d6f2d7c67e[…]a108b2a/sds_data_manager/lambda_code/SDSCode/database/models.py](https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/d6f2d7c67ed7f1acb3d11b9f7ff77030ba108b2a/sds_data_manager/lambda_code/SDSCode/database/models.py#L113-L118)
* With the unique constraint in place, switch to a try/except around writing to the DB. A write will be guaranteed to fail if a record already exists and we can catch that exception and know a job already exists. This will avoid the time delay between query / write.
* Move to a queue-based job starter system that 
[@Maxine](https://lasp.slack.com/team/UC2209B41)
 is working on.

I think all of the first 3 bullets are relatively minor updates and we should probably add them in regardless of the queue system as a backup and best practice, unless the queue system removes the DB work entirely Maxine?
[database_handler.py](https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/d6f2d7c67ed7f1acb3d11b9f7ff77030ba108b2a/sds_data_manager/lambda_code/SDSCode/database_handler.py)
<https://github.com/[IMAP-Science-Operations-Center/sds-data-manager](https://github.com/IMAP-Science-Operations-Center/sds-data-manager)|IMAP-Science-Operations-Center/sds-data-manager>IMAP-Science-Operations-Center/sds-data-manager | Added by [GitHub](https://lasp.slack.com/services/B02RT0NPJNL)
[models.py](https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/d6f2d7c67ed7f1acb3d11b9f7ff77030ba108b2a/sds_data_manager/lambda_code/SDSCode/database/models.py)
<https://github.com/[IMAP-Science-Operations-Center/sds-data-manager](https://github.com/IMAP-Science-Operations-Center/sds-data-manager)|IMAP-Science-Operations-Center/sds-data-manager>IMAP-Science-Operations-Center/sds-data-manager | Added by [GitHub](https://lasp.slack.com/services/B02RT0NPJNL)

--- a/sds_data_manager/lambda_code/SDSCode/batch_starter.py
+++ b/sds_data_manager/lambda_code/SDSCode/batch_starter.py
@@ -7,20 +7,23 @@ from datetime import datetime
 import boto3
 from imap_data_access import ScienceFilePath
 from sqlalchemy import select
+from sqlalchemy.exc import IntegrityError
 
 from .database import database as db
 from .database import models
-from .database_handler import update_status_table
 
 # Logger setup
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
 
+# Create a batch client
+BATCH_CLIENT = boto3.client("batch", region_name="us-west-2")
 
-def get_dependency(
+
+def get_dependencies(
     session, instrument, data_level, descriptor, direction, relationship
 ):
-    """Make query to dependency table to get dependency.
+    """Make a query to the dependency table to get all dependencies.
 
     Parameters
     ----------
@@ -40,10 +43,10 @@ def get_dependency(
 
     Returns
     -------
-    dependency : list
+    dependencies : list
         List of dictionary containing the dependency information.
     """
-    dependency = []
+    dependencies = []
 
     query = select(models.PreProcessingDependency.__table__).where(
         models.PreProcessingDependency.primary_instrument == instrument,
@@ -55,25 +58,29 @@ def get_dependency(
     results = session.execute(query).all()
 
     for result in results:
-        dependency.append(
+        dependencies.append(
             {
                 "instrument": result.dependent_instrument,
                 "data_level": result.dependent_data_level,
                 "descriptor": result.dependent_descriptor,
             }
         )
-    return dependency
+    return dependencies
 
 
-def query_instrument(session, upstream_dependency, start_date, version):
+def get_file(session, instrument, data_level, descriptor, start_date, version):
     """Query to database to get the first FileCatalog record.
 
     Parameters
     ----------
     session : orm session
         Database session.
-    upstream_dependency : dict
-        Dictionary of upstream dependency.
+    instrument : str
+        Instrument name.
+    data_level : str
+        Data level.
+    descriptor : str
+        Data descriptor.
     start_date : str
         Start date of the event data.
     version : str
@@ -85,10 +92,6 @@ def query_instrument(session, upstream_dependency, start_date, version):
         The first FileCatalog record matching the query criteria.
         None is returned if no record matches the criteria.
     """
-    instrument = upstream_dependency["instrument"]
-    data_level = upstream_dependency["data_level"]
-    descriptor = upstream_dependency["descriptor"]
-
     # TODO: narrow down the query using end_date.
     # This will give ability to query range of time.
     # Eg. when we are query 3 months of data to create
@@ -101,9 +104,9 @@ def query_instrument(session, upstream_dependency, start_date, version):
         .filter(
             models.FileCatalog.instrument == instrument,
             models.FileCatalog.data_level == data_level,
-            models.FileCatalog.version == version,
             models.FileCatalog.descriptor == descriptor,
             models.FileCatalog.start_date == datetime.strptime(start_date, "%Y%m%d"),
+            models.FileCatalog.version == version,
         )
         .first()
     )
@@ -111,7 +114,7 @@ def query_instrument(session, upstream_dependency, start_date, version):
     return record
 
 
-def query_downstream_dependencies(session, filename_components):
+def get_downstream_dependencies(session, filename_components):
     """Get information of downstream dependents.
 
     Parameters
@@ -127,7 +130,7 @@ def query_downstream_dependencies(session, filename_components):
         Dictionary containing components with dates and versions appended.
     """
     # Get downstream dependency data
-    downstream_dependents = get_dependency(
+    downstream_dependents = get_dependencies(
         session=session,
         instrument=filename_components["instrument"],
         data_level=filename_components["data_level"],
@@ -149,129 +152,7 @@ def query_downstream_dependencies(session, filename_components):
     return downstream_dependents
 
 
-def query_upstream_dependencies(session, downstream_dependents):
-    """Find dependency information for each instrument.
-
-    This function looks for upstream dependency of current downstream
-    dependents.
-
-    Parameters
-    ----------
-    session : orm session
-        Database session.
-    downstream_dependents : list of dict
-        Dictionary containing components with dates and versions appended.
-
-    Returns
-    -------
-    instruments_to_process : list of dict
-        A list of dictionaries containing the information needed for batch
-        job command.
-    """
-    instruments_to_process = []
-
-    # Iterate over each downstream dependent
-    for dependent in downstream_dependents:
-        instrument = dependent["instrument"]
-        data_level = dependent["data_level"]
-        descriptor = dependent["descriptor"]
-        start_date = dependent["start_date"]
-        version = dependent["version"]
-
-        logger.info("Checking for job in progress before looking for dependencies.")
-
-        job_already_exist = is_job_in_status_table(
-            session=session,
-            instrument=instrument,
-            data_level=data_level,
-            descriptor=descriptor,
-            start_date=start_date,
-            version=version,
-        )
-
-        if job_already_exist:
-            logger.info(
-                f"Job already in progress for {instrument}, {data_level}, "
-                f"{descriptor}, {start_date}, {version}"
-            )
-            continue
-
-        logger.info(
-            f"Job not in progress for {instrument}, {data_level}, "
-            f"{descriptor}, {start_date}, {version}"
-        )
-
-        # For each downstream dependent, find its upstream dependencies
-        upstream_dependencies = get_dependency(
-            session=session,
-            instrument=instrument,
-            data_level=data_level,
-            descriptor=descriptor,
-            direction="UPSTREAM",
-            relationship="HARD",
-        )
-
-        all_dependencies_available = True  # Initialize the flag
-        for upstream_dependency in upstream_dependencies:
-            # Check to see if each upstream dependency is available
-            record = query_instrument(session, upstream_dependency, start_date, version)
-            if record is None:
-                all_dependencies_available = (
-                    False  # Set flag to false if any dependency is missing
-                )
-                logger.info(
-                    f"Missing dependency: {upstream_dependency['instrument']}, "
-                    f"{upstream_dependency['data_level']}, "
-                    f"{upstream_dependency['descriptor']}, "
-                    f"{version}"
-                )
-                logger.info(
-                    f"Failed to kickoff this downstream dependent: {dependent}, "
-                    "because it requires data for all these upstream "
-                    f" dependencies: {upstream_dependencies}"
-                )
-                break  # Exit the loop early as we already found a missing dependency
-            else:
-                logger.info(
-                    f"Dependency found: {upstream_dependency['instrument']}, "
-                    f"{upstream_dependency['data_level']}, "
-                    f"{upstream_dependency['descriptor']}, "
-                    f"{version}"
-                )
-                # Add additional information to the upstream dependency
-                additional_info = {
-                    "start_date": start_date,
-                    "version": version,
-                }
-                upstream_dependency.update(additional_info)
-
-        # If all dependencies are available, prepare the data for batch job
-        if all_dependencies_available:
-            # Add it's information to the instrument to process list
-            instruments_to_process.append(
-                {
-                    "instrument": instrument,
-                    "data_level": data_level,
-                    "descriptor": descriptor,
-                    "start_date": start_date,
-                    "version": version,
-                    "upstream_dependencies": upstream_dependencies,
-                }
-            )
-
-            logger.info(f"All dependencies for {instrument} present.")
-            logger.info(
-                f"For this downstream dependent: {dependent}, "
-                "all these upstream dependencies data are available: "
-                f"{upstream_dependencies}"
-            )
-        else:
-            logger.info(f"Some dependencies for {instrument} are missing.")
-
-    return instruments_to_process
-
-
-def is_job_in_status_table(
+def is_job_in_processing_table(
     session: db.Session,
     instrument: str,
     data_level: str,
@@ -301,15 +182,15 @@ def is_job_in_status_table(
     bool
         True if duplicate job is found, False otherwise.
     """
-    # check in status tracking table if job is already in progress
+    # check in the processing table if the job is already in progress
     # for this instrument, data level, version, and descriptor
-    query = select(models.StatusTracking.__table__).where(
-        models.StatusTracking.instrument == instrument,
-        models.StatusTracking.data_level == data_level,
-        models.StatusTracking.descriptor == descriptor,
-        models.StatusTracking.start_date == datetime.strptime(start_date, "%Y%m%d"),
-        models.StatusTracking.version == version,
-        models.StatusTracking.status.in_(
+    query = select(models.ProcessingJob.__table__).where(
+        models.ProcessingJob.instrument == instrument,
+        models.ProcessingJob.data_level == data_level,
+        models.ProcessingJob.descriptor == descriptor,
+        models.ProcessingJob.start_date == datetime.strptime(start_date, "%Y%m%d"),
+        models.ProcessingJob.version == version,
+        models.ProcessingJob.status.in_(
             [models.Status.INPROGRESS.value, models.Status.SUCCEEDED.value]
         ),
     )
@@ -320,23 +201,146 @@ def is_job_in_status_table(
     return False
 
 
-def lambda_handler(event: dict, context):
-    """Lambda handler."""
-    logger.info(f"Event: {event}")
-    logger.info(f"Context: {context}")
+def try_to_submit_job(session, job_info):
+    """Try to submit a batch job with the given job information.
 
-    # Event details:
-    filename = event["detail"]["object"]["key"]
-    components = ScienceFilePath.extract_filename_components(filename)
-    logger.info(f"Parsed filename - {components}")
-    instrument = components["instrument"]
+    Go through the job information to retrieve all necessary input files
+    (upstream dependencies). If any are missing, return. If we have
+    all the necessary input files, submit the job to the batch queue.
 
-    # Get information for the batch job.
+    Parameters
+    ----------
+    session : orm session
+        Database session.
+    job_info : dict
+        Dictionary containing components with dates and versions appended.
+
+    Returns
+    -------
+    bool
+        Whether or not this job is ready to be processed.
+    """
+    instrument = job_info["instrument"]
+    data_level = job_info["data_level"]
+    descriptor = job_info["descriptor"]
+    start_date = job_info["start_date"]
+    version = job_info["version"]
+
+    logger.info("Checking for job in progress before looking for dependencies.")
+
+    if is_job_in_processing_table(
+        session=session,
+        instrument=instrument,
+        data_level=data_level,
+        descriptor=descriptor,
+        start_date=start_date,
+        version=version,
+    ):
+        logger.info(
+            f"Job already in progress for {instrument}, {data_level}, "
+            f"{descriptor}, {start_date}, {version}"
+        )
+        return
+
+    # Find the files that this job depends on
+    upstream_dependencies = get_dependencies(
+        session=session,
+        instrument=instrument,
+        data_level=data_level,
+        descriptor=descriptor,
+        direction="UPSTREAM",
+        relationship="HARD",
+    )
+
+    for upstream_dependency in upstream_dependencies:
+        upstream_instrument = upstream_dependency["instrument"]
+        upstream_data_level = upstream_dependency["data_level"]
+        upstream_descriptor = upstream_dependency["descriptor"]
+
+        # Check to see if each upstream dependency file is available
+        # TODO: Update start_date / version request to be more specific
+        #       Currently we are using the same as the job product, but the versions
+        #       may not match exactly if one dependency updates before another
+        upstream_start_date = start_date
+        upstream_version = version
+        upstream_dependency.update(
+            {"start_date": upstream_start_date, "version": upstream_version}
+        )
+        record = get_file(
+            session,
+            upstream_instrument,
+            upstream_data_level,
+            upstream_descriptor,
+            upstream_start_date,
+            upstream_version,
+        )
+        if record is None:
+            logger.info(
+                f"Dependency not found: {upstream_instrument}, "
+                f"{upstream_data_level}, "
+                f"{upstream_descriptor}, "
+                f"{upstream_start_date}, "
+                f"{upstream_version}"
+            )
+            return  # Exit the loop early as we already found a missing dependency
+    logger.info(f"All dependencies found for the job: {job_info}")
+
+    # All of our upstream requirements have been met.
+    # Try to insert a record into the Processing Jobs table
+    # If this job already exists, then we will get an integrity error
+    # and know that some other process has already taken care of it
+    processing_job = models.ProcessingJob(
+        status=models.Status.INPROGRESS,
+        instrument=instrument,
+        data_level=data_level,
+        descriptor=descriptor,
+        start_date=datetime.strptime(start_date, "%Y%m%d"),
+        version=version,
+    )
+
+    try:
+        session.add(processing_job)
+        session.commit()
+    except IntegrityError:
+        logger.info(f"Job already completed or in progress: {processing_job}")
+        return
+
+    logger.info(
+        f"Wrote job INPROGRESS to Processing Jobs Table with id: {processing_job.id}"
+    )
+
+    # FYI, these are the keys the upstream_dependencies should contain:
+    # {
+    #     'instrument': 'swe',
+    #     'data_level': 'l0',
+    #     'descriptor': 'lveng-hk',
+    #     'start_date': '20231212',
+    #     'version': 'v001',
+    # },
+    batch_command = [
+        "--instrument",
+        instrument,
+        "--data-level",
+        data_level,
+        "--descriptor",
+        descriptor,
+        "--start-date",
+        start_date,
+        "--version",
+        version,
+        "--dependency",
+        f"{upstream_dependencies}",
+        "--upload-to-sdc",
+    ]
+
+    # NOTE: The batch job name should contain only alphanumeric characters and hyphens
+    # Eg. "codice-l1a-sci-job-1"
+    # The `processing_job.id` is used later for updating the job processing table
+    job_name = f"{instrument}-{data_level}-{descriptor}-job-{processing_job.id}"
+    # Get the necessary AWS information
+    # NOTE: These are here for easier mocking in tests rather than at the module level
     region = os.getenv("REGION")
     account = os.getenv("ACCOUNT")
-    # Create a batch client
-    batch_client = boto3.client("batch")
-
     job_definition = (
         f"arn:aws:batch:{region}:{account}:job-definition/"
         f"fargate-batch-job-definition{instrument}"
@@ -345,113 +349,31 @@ def lambda_handler(event: dict, context):
         f"arn:aws:batch:{region}:{account}:job-queue/"
         f"{instrument}-fargate-batch-job-queue"
     )
+    BATCH_CLIENT.submit_job(
+        jobName=job_name,
+        jobQueue=job_queue,
+        jobDefinition=job_definition,
+        containerOverrides={
+            "command": batch_command,
+        },
+    )
+    logger.info(f"Submitted job {job_name} with this command: {batch_command}")
+
+
+def lambda_handler(event: dict, context):
+    """Lambda handler."""
+    logger.info(f"Event: {event}")
+    logger.info(f"Context: {context}")
+
+    # Event details:
+    filename = event["detail"]["object"]["key"]
+    components = ScienceFilePath.extract_filename_components(filename)
+    logger.info(f"Initial event parsed filename: {components}")
 
     with db.Session() as session:
-        # Downstream dependents are the instruments that
-        # depend on the current instrument.
-        downstream_dependents = query_downstream_dependencies(session, components)
+        # Potential jobs are the instruments that depend on the current file.
+        potential_jobs = get_downstream_dependencies(session, components)
+        logger.info(f"Potential jobs found [{len(potential_jobs)}]: {potential_jobs}")
 
-        # Check if every downstream dependents
-        # have all upstream dependencies. This helps to determine if
-        # we can start the batch job.
-        downstream_instruments_to_process = query_upstream_dependencies(
-            session, downstream_dependents
-        )
-
-        # No instruments to process
-        if not downstream_instruments_to_process:
-            logger.info("No instruments_to_process. Skipping further processing.")
-            return
-
-        # Start Batch Job execution for those that has all dependencies
-        for downstream_data in downstream_instruments_to_process:
-            logger.info("Checking for duplicate job before kicking off the job.")
-
-            job_already_exist = is_job_in_status_table(
-                session=session,
-                instrument=downstream_data["instrument"],
-                data_level=downstream_data["data_level"],
-                descriptor=downstream_data["descriptor"],
-                start_date=downstream_data["start_date"],
-                version=downstream_data["version"],
-            )
-
-            if job_already_exist:
-                logger.info(
-                    f"Job record found for {downstream_data['instrument']}, "
-                    f"{downstream_data['data_level']}, "
-                    f"{downstream_data['descriptor']}, "
-                    f"{downstream_data['start_date']}, "
-                    f"{downstream_data['version']}"
-                )
-                continue
-
-            logger.info(
-                f"Job record not found for {downstream_data['instrument']}, "
-                f"{downstream_data['data_level']}, "
-                f"{downstream_data['descriptor']}, "
-                f"{downstream_data['start_date']}, "
-                f"{downstream_data['version']}"
-            )
-
-            # Write to status tracking table with initial values
-            status_params = {
-                "status": models.Status.INPROGRESS,
-                "instrument": downstream_data["instrument"],
-                "data_level": downstream_data["data_level"],
-                "descriptor": downstream_data["descriptor"],
-                "start_date": datetime.strptime(
-                    downstream_data["start_date"], "%Y%m%d"
-                ),
-                "version": downstream_data["version"],
-            }
-
-            update_status_table(session, status_params)
-
-            logger.info(
-                f"Wrote job in progress to status tracking table - {status_params}"
-            )
-
-            # FYI, these are the keys the upstream_dependencies should contain:
-            # {
-            #     'instrument': 'swe',
-            #     'data_level': 'l0',
-            #     'descriptor': 'lveng-hk',
-            #     'start_date': '20231212',
-            #     'version': 'v001',
-            # },
-            batch_command = [
-                "--instrument",
-                downstream_data["instrument"],
-                "--data-level",
-                downstream_data["data_level"],
-                "--descriptor",
-                downstream_data["descriptor"],
-                "--start-date",
-                downstream_data["start_date"],
-                "--version",
-                downstream_data["version"],
-                "--dependency",
-                f"{downstream_data['upstream_dependencies']}",
-                "--upload-to-sdc",
-            ]
-
-            logger.info(f"Submitting job with this command - {batch_command}")
-            # NOTE: The batch job name should contain only
-            # alphanumeric characters and hyphens.
-            level_to_process = downstream_data["data_level"]
-            # Eg. "codice-l1a-job"
-            job_name = (
-                f"{downstream_data['instrument']}-{level_to_process}-"
-                f"{downstream_data['descriptor']}-job"
-            )
-            logger.info("Job name: %s", job_name)
-            response = batch_client.submit_job(
-                jobName=job_name,
-                jobQueue=job_queue,
-                jobDefinition=job_definition,
-                containerOverrides={
-                    "command": batch_command,
-                },
-            )
-            logger.info(f"Submitted job - {response}")
+        for job in potential_jobs:
+            try_to_submit_job(session, job)

--- a/sds_data_manager/lambda_code/SDSCode/database/models.py
+++ b/sds_data_manager/lambda_code/SDSCode/database/models.py
@@ -8,7 +8,6 @@ from enum import Enum
 
 from sqlalchemy import (
     Boolean,
-    CheckConstraint,
     Column,
     DateTime,
     Identity,
@@ -108,10 +107,10 @@ class UniversalSpinTable(Base):
     repointing_number = Column(Integer, nullable=False)
 
 
-class StatusTracking(Base):
-    """Status tracking table."""
+class ProcessingJob(Base):
+    """Track all processing jobs."""
 
-    __tablename__ = "status_tracking"
+    __tablename__ = "processing_job_table"
 
     id = Column(Integer, Identity(start=1, increment=1), primary_key=True)
     status = Column(STATUSES, nullable=False)
@@ -142,11 +141,6 @@ class StatusTracking(Base):
             "version",
             unique=True,
             postgresql_where=and_(status.in_(["INPROGRESS", "SUCCEEDED"])),
-        ),
-        # Optional: Check constraint to ensure valid status values
-        CheckConstraint(
-            status.in_(["INPROGRESS", "SUCCEEDED", "FAILED"]),
-            name="check_status_values",
         ),
     )
 

--- a/sds_data_manager/lambda_code/SDSCode/database_handler.py
+++ b/sds_data_manager/lambda_code/SDSCode/database_handler.py
@@ -11,59 +11,6 @@ logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
 
 
-def update_status_table(session, status_params):
-    """Update status tracking table.
-
-    Parameters
-    ----------
-    session : sqlalchemy.orm.session.Session
-        Database session.
-    status_params : dict
-        Data information
-
-    """
-    instrument = status_params["instrument"]
-    data_level = status_params["data_level"]
-    descriptor = status_params["descriptor"]
-    start_date = status_params["start_date"]
-    version = status_params["version"]
-
-    try:
-        # Had to query this way because the select statement
-        # returns a RowProxy object when it executes it,
-        # not the actual StatusTracking model instance,
-        # which is why it can't update table row directly.
-        result = (
-            session.query(models.StatusTracking)
-            .filter(models.StatusTracking.instrument == instrument)
-            .filter(models.StatusTracking.data_level == data_level)
-            .filter(models.StatusTracking.descriptor == descriptor)
-            .filter(models.StatusTracking.start_date == start_date)
-            .filter(models.StatusTracking.version == version)
-            .first()
-        )
-
-        if result is None:
-            logger.info(
-                "No existing record found, creating"
-                f" new record for {instrument}, {data_level},"
-                f"{descriptor}, {start_date}, {version}"
-            )
-            session.add(models.StatusTracking(**status_params))
-            session.commit()
-        else:
-            logger.info(f"Updating existing record, {result.__dict__}, with batch info")
-            result.status = status_params["status"]
-            result.job_definition = status_params["job_definition"]
-            result.job_log_stream_id = status_params["job_log_stream_id"]
-            result.container_image = status_params["container_image"]
-            result.container_command = status_params["container_command"]
-            session.commit()
-
-    except IntegrityError as e:
-        logger.error(str(e))
-
-
 def update_file_catalog_table(session, metadata_params):
     """Update file catalog table.
 

--- a/sds_data_manager/lambda_code/SDSCode/indexer.py
+++ b/sds_data_manager/lambda_code/SDSCode/indexer.py
@@ -10,7 +10,7 @@ from imap_data_access import ScienceFilePath
 
 from .database import database as db
 from .database import models
-from .database_handler import update_file_catalog_table, update_status_table
+from .database_handler import update_file_catalog_table
 from .lambda_custom_events import IMAPLambdaPutEvent
 
 # Logger setup
@@ -250,15 +250,6 @@ def batch_event_handler(event):
         HTTP response
 
     """
-    command = event["detail"]["container"]["command"]
-
-    # Get params from batch job command
-    instrument = command[1]
-    data_level = command[3]
-    descriptor = command[5]
-    start_date = datetime.strptime(command[7], "%Y%m%d")
-    version = command[9]
-
     # Get job status
     job_status = (
         models.Status.SUCCEEDED
@@ -266,21 +257,19 @@ def batch_event_handler(event):
         else models.Status.FAILED
     )
 
-    status_params = {
-        "status": job_status,
-        "instrument": instrument,
-        "data_level": data_level,
-        "descriptor": descriptor,
-        "start_date": start_date,
-        "version": version,
-        "job_definition": event["detail"]["jobDefinition"],
-        "job_log_stream_id": event["detail"]["container"]["logStreamName"],
-        "container_image": event["detail"]["container"]["image"],
-        "container_command": " ".join(command),
-    }
+    # We injected our table ID into the job name
+    job_id = event["detail"]["jobName"].split("-")[-1]
 
     with db.Session() as session:
-        update_status_table(session, status_params)
+        # Get the batch job by its ID
+        job = session.get(models.ProcessingJob, job_id)
+        # Make the updates
+        job.status = job_status
+        job.job_definition = event["detail"]["jobDefinition"]
+        job.job_log_stream_id = event["detail"]["container"]["logStreamName"]
+        job.container_image = event["detail"]["container"]["image"]
+        job.container_command = " ".join(event["detail"]["container"]["command"])
+        session.commit()
 
     return http_response(status_code=200, body="Success")
 

--- a/tests/lambda_endpoints/conftest.py
+++ b/tests/lambda_endpoints/conftest.py
@@ -5,7 +5,7 @@ from unittest.mock import patch
 
 import boto3
 import pytest
-from moto import mock_batch, mock_events, mock_s3
+from moto import mock_events, mock_s3
 from sqlalchemy import create_engine
 from sqlalchemy.orm import sessionmaker
 
@@ -37,13 +37,6 @@ def science_file():
 def spice_file():
     """Path to a valid spice file."""
     return "imap/spice/ck/test_v000.bc"
-
-
-@pytest.fixture()
-def batch_client():
-    """Yield a batch client."""
-    with mock_batch():
-        yield boto3.client("batch", region_name="us-west-2")
 
 
 @pytest.fixture(autouse=True, scope="module")

--- a/tests/lambda_endpoints/test_batch_starter.py
+++ b/tests/lambda_endpoints/test_batch_starter.py
@@ -2,23 +2,24 @@
 
 import copy
 from datetime import datetime
+from unittest.mock import Mock, patch
 
 import pytest
 from imap_data_access import ScienceFilePath
 from sqlalchemy.exc import IntegrityError
 
+from sds_data_manager.lambda_code.SDSCode import batch_starter
 from sds_data_manager.lambda_code.SDSCode.batch_starter import (
-    get_dependency,
-    is_job_in_status_table,
+    get_dependencies,
+    get_downstream_dependencies,
+    get_file,
+    is_job_in_processing_table,
     lambda_handler,
-    query_downstream_dependencies,
-    query_instrument,
-    query_upstream_dependencies,
 )
 from sds_data_manager.lambda_code.SDSCode.database import models
 from sds_data_manager.lambda_code.SDSCode.database.models import (
     FileCatalog,
-    StatusTracking,
+    ProcessingJob,
 )
 from sds_data_manager.lambda_code.SDSCode.dependency_config import (
     all_dependents,
@@ -121,12 +122,11 @@ def _populate_file_catalog(session):
     session.commit()
 
 
-def _populate_status_tracking_table(session):
+def _populate_processing_table(session):
     """Add test data to database."""
-    # Add an inprogress record to the status_tracking table
-    # to test the is_job_in_status_table function.
+    # Add an inprogress record to the processing table
     # At the time of job kickoff, we only have these written to the table
-    record = StatusTracking(
+    record = ProcessingJob(
         status=models.Status.INPROGRESS,
         instrument="lo",
         data_level="l1b",
@@ -210,7 +210,7 @@ def test_pre_processing_dependency(session):
     """Test pre-processing dependency."""
     _populate_dependency_table(session)
     # upstream dependency
-    upstream_dependency = get_dependency(
+    upstream_dependency = get_dependencies(
         session=session,
         instrument="mag",
         data_level="l1a",
@@ -224,7 +224,7 @@ def test_pre_processing_dependency(session):
     assert upstream_dependency[0]["descriptor"] == "raw"
 
     # downstream dependency
-    downstream_dependency = get_dependency(
+    downstream_dependency = get_dependencies(
         session=session,
         instrument="mag",
         data_level="l1b",
@@ -238,37 +238,44 @@ def test_pre_processing_dependency(session):
     assert downstream_dependency[0]["descriptor"] == "norm-mago"
 
 
-def test_query_instrument(session):
-    """Tests ``query_instrument`` function."""
+def test_get_file(session):
+    """Tests the get_file function."""
     _populate_file_catalog(session)
-    upstream_dependency = {
-        "instrument": "ultra",
-        "data_level": "l2",
-        "version": "v001",
-        "descriptor": "sci",
-    }
 
-    "Tests query_instrument function."
-    record = query_instrument(
+    record = get_file(
         session,
-        upstream_dependency,
-        "20240101",
-        "v001",
+        instrument="ultra",
+        data_level="l2",
+        descriptor="sci",
+        start_date="20240101",
+        version="v001",
     )
 
     assert record.instrument == "ultra"
     assert record.data_level == "l2"
-    assert record.version == "v001"
+    assert record.descriptor == "sci"
     assert record.start_date == datetime(2024, 1, 1)
+    assert record.version == "v001"
+
+    # Non-existent record should return None
+    record = get_file(
+        session,
+        instrument="ultra",
+        data_level="l2",
+        descriptor="sci",
+        start_date="20000101",
+        version="v001",
+    )
+    assert record is None
 
 
-def test_query_downstream_dependencies(session):
-    "Tests query_downstream_dependencies function."
+def test_get_downstream_dependencies(session):
+    "Tests get_downstream_dependencies function."
     _populate_dependency_table(session)
     filename = "imap_hit_l1a_sci_20240101_v001.cdf"
     file_params = ScienceFilePath.extract_filename_components(filename)
 
-    complete_dependents = query_downstream_dependencies(session, file_params)
+    complete_dependents = get_downstream_dependencies(session, file_params)
     expected_complete_dependent = {
         "instrument": "hit",
         "data_level": "l1b",
@@ -276,119 +283,36 @@ def test_query_downstream_dependencies(session):
         "version": "v001",
         "start_date": "20240101",
     }
+    assert len(complete_dependents) == 1
 
     assert complete_dependents[0] == expected_complete_dependent
 
 
-def test_query_upstream_dependencies(session):
-    """Tests ``query_upstream_dependencies`` function."""
-    _populate_dependency_table(session)
-    _populate_file_catalog(session)
-    downstream_dependents = [
-        {
-            "instrument": "hit",
-            "data_level": "l1a",
-            "version": "v001",
-            "descriptor": "sci",
-            "start_date": "20240101",
-        },
-        {
-            "instrument": "hit",
-            "data_level": "l3",
-            "version": "v001",
-            "descriptor": "sci",
-            "start_date": "20240101",
-        },
-    ]
-
-    result = query_upstream_dependencies(session, downstream_dependents)
-    assert result[0]["instrument"] == "hit"
-    assert result[0]["data_level"] == "l1a"
-    assert result[0]["version"] == "v001"
-    expected_upstream_dependents = [
-        {
-            "instrument": "hit",
-            "data_level": "l0",
-            "descriptor": "sci",
-            "start_date": "20240101",
-            "version": "v001",
-        }
-    ]
-    assert result[0]["upstream_dependencies"] == expected_upstream_dependents
-
-    # find swe upstream dependencies
-    downstream_dependents = [
-        {
-            "instrument": "swe",
-            "data_level": "l1a",
-            "version": "v001",
-            "descriptor": "sci",
-            "start_date": "20240101",
-        }
-    ]
-
-    result = query_upstream_dependencies(session, downstream_dependents)
-
-    assert len(result) == 1
-    assert result[0]["instrument"] == "swe"
-    assert result[0]["data_level"] == "l1a"
-    assert result[0]["version"] == "v001"
-    expected_upstream_dependents = [
-        {
-            "instrument": "swe",
-            "data_level": "l0",
-            "descriptor": "raw",
-            "start_date": "20240101",
-            "version": "v001",
-        }
-    ]
-
-    assert result[0]["upstream_dependencies"] == expected_upstream_dependents
-
-    downstream_dependents = [
-        {
-            "instrument": "swe",
-            "data_level": "l1b",
-            "version": "v001",
-            "descriptor": "sci",
-            "start_date": "20240101",
-        }
-    ]
-
-    result = query_upstream_dependencies(session, downstream_dependents)
-
-    assert len(result) == 1
-    assert result[0]["instrument"] == "swe"
-    assert result[0]["data_level"] == "l1b"
-    assert result[0]["version"] == "v001"
-    expected_upstream_dependents = [
-        {
-            "instrument": "swe",
-            "data_level": "l1a",
-            "descriptor": "sci",
-            "start_date": "20240101",
-            "version": "v001",
-        }
-    ]
-
-    assert result[0]["upstream_dependencies"] == expected_upstream_dependents
-
-
 def test_lambda_handler(
     session,
-    batch_client,
 ):
     """Tests ``lambda_handler`` function."""
-    event = {"detail": {"object": {"key": "imap_hit_l1a_sci_20100101_v001.cdf"}}}
+    _populate_dependency_table(session)
+    _populate_file_catalog(session)
+
+    event = {"detail": {"object": {"key": "imap_swe_l0_raw_20240101_v001.pkts"}}}
     context = {"context": "sample_context"}
-    lambda_handler(event, context)
+    with patch.object(batch_starter, "BATCH_CLIENT", Mock()) as mock_batch_client:
+        lambda_handler(event, context)
+        mock_batch_client.submit_job.assert_called_once()
+
+        # Submit a second job with the same file as input which will try to kick
+        # off a duplicate job. We expect the submit_job method to not be called
+        # so make sure it is still only called once from our previous iteration.
+        lambda_handler(event, context)
+        mock_batch_client.submit_job.assert_called_once()
 
 
 def test_is_job_in_status_table(session):
     """Test the ``is_job_in_status_table`` function."""
-    _populate_status_tracking_table(session)
-    # query the status_tracking table if this job is already in progress
-    result = is_job_in_status_table(
+    _populate_processing_table(session)
+    # query the processing table if this job is already in progress
+    result = is_job_in_processing_table(
         session=session,
         instrument="lo",
         data_level="l1b",
@@ -399,7 +323,7 @@ def test_is_job_in_status_table(session):
 
     assert result
 
-    result = is_job_in_status_table(
+    result = is_job_in_processing_table(
         session=session,
         instrument="swapi",
         data_level="l1b",
@@ -422,11 +346,11 @@ def test_is_job_in_status_table(session):
 )
 def test_duplicate_job(session, first_status, second_status):
     """Multiple jobs in progress should raise an IntegrityError."""
-    # Add some initial FAILED entries to the status_tracking table
+    # Add some initial FAILED entries to the processing table
     # These should not be a part of the unique constraint
     for _ in range(3):
         session.add(
-            StatusTracking(
+            ProcessingJob(
                 status=models.Status.FAILED,
                 instrument="lo",
                 data_level="l1b",
@@ -436,9 +360,9 @@ def test_duplicate_job(session, first_status, second_status):
             )
         )
     session.commit()
-    assert session.query(StatusTracking).count() == 3
+    assert session.query(ProcessingJob).count() == 3
 
-    record = StatusTracking(
+    record = ProcessingJob(
         status=first_status,
         instrument="lo",
         data_level="l1b",
@@ -448,9 +372,9 @@ def test_duplicate_job(session, first_status, second_status):
     )
     session.add(record)
     session.commit()
-    assert session.query(StatusTracking).count() == 4
+    assert session.query(ProcessingJob).count() == 4
 
-    duplicate = StatusTracking(
+    duplicate = ProcessingJob(
         status=second_status,
         instrument="lo",
         data_level="l1b",
@@ -465,10 +389,10 @@ def test_duplicate_job(session, first_status, second_status):
     session.rollback()
 
     # Now we should still only have 4 items in the table
-    assert session.query(StatusTracking).count() == 4
+    assert session.query(ProcessingJob).count() == 4
 
     # We can add another FAILED status without issue
-    record = StatusTracking(
+    record = ProcessingJob(
         status=models.Status.FAILED,
         instrument="lo",
         data_level="l1b",
@@ -478,4 +402,4 @@ def test_duplicate_job(session, first_status, second_status):
     )
     session.add(record)
     session.commit()
-    assert session.query(StatusTracking).count() == 5
+    assert session.query(ProcessingJob).count() == 5

--- a/tests/lambda_endpoints/test_indexer_lambda.py
+++ b/tests/lambda_endpoints/test_indexer_lambda.py
@@ -9,9 +9,6 @@ from sqlalchemy import select
 
 from sds_data_manager.lambda_code.SDSCode import indexer
 from sds_data_manager.lambda_code.SDSCode.database import models
-from sds_data_manager.lambda_code.SDSCode.database_handler import (
-    update_status_table,
-)
 from sds_data_manager.lambda_code.SDSCode.indexer import (
     send_event_from_indexer,
 )
@@ -19,8 +16,8 @@ from sds_data_manager.lambda_code.SDSCode.indexer import (
 
 def test_batch_job_event(session, events_client):
     """Test batch job event."""
-    # Write to status tracking table with current batch job event info
-    status_params = {
+    # Write to Processing job table with current batch job event info
+    job_params = {
         "status": models.Status.INPROGRESS,
         "instrument": "swapi",
         "data_level": "l1",
@@ -28,7 +25,10 @@ def test_batch_job_event(session, events_client):
         "start_date": datetime.strptime("20230724", "%Y%m%d"),
         "version": "v001",
     }
-    update_status_table(session, status_params)
+    processing_job = models.ProcessingJob(**job_params)
+    session.add(processing_job)
+    session.commit()
+    job_id = processing_job.id
 
     # TODO: Will update this test further
     # when I extend batch job event handler.
@@ -40,7 +40,7 @@ def test_batch_job_event(session, events_client):
                 "arn:aws:batch:us-west-2:012345678910:"
                 "job/26242c7e-3d49-4e41-9387-74fcaf9630bb"
             ),
-            "jobName": "swe-l0-job",
+            "jobName": f"swe-l0-job-{job_id}",  # NOTE: We need to add job_id to jobName
             "jobId": "26242c7e-3d49-4e41-9387-74fcaf9630bb",
             "jobQueue": (
                 "arn:aws:batch:us-west-2:012345678910:"
@@ -89,40 +89,30 @@ def test_batch_job_event(session, events_client):
     assert returned_value["statusCode"] == 200
 
     # check that data was written to status table
-    query = select(models.StatusTracking.__table__).where(
-        models.StatusTracking.instrument == status_params["instrument"],
-        models.StatusTracking.data_level == status_params["data_level"],
-        models.StatusTracking.version == status_params["version"],
+    query = select(models.ProcessingJob.__table__).where(
+        models.ProcessingJob.instrument == job_params["instrument"],
+        models.ProcessingJob.data_level == job_params["data_level"],
+        models.ProcessingJob.version == job_params["version"],
     )
 
-    status_tracking = session.execute(query).first()
-    assert status_tracking.status == models.Status.FAILED
+    processing_job = session.execute(query).first()
+    assert processing_job.id == job_id
+    assert processing_job.status == models.Status.FAILED
 
     # Test for succeeded case
     event["detail"]["status"] = "SUCCEEDED"
     returned_value = indexer.lambda_handler(event=event, context={})
     assert returned_value["statusCode"] == 200
 
-    query = select(models.StatusTracking.__table__).where(
-        models.StatusTracking.instrument == status_params["instrument"],
-        models.StatusTracking.data_level == status_params["data_level"],
-        models.StatusTracking.version == status_params["version"],
+    query = select(models.ProcessingJob.__table__).where(
+        models.ProcessingJob.instrument == job_params["instrument"],
+        models.ProcessingJob.data_level == job_params["data_level"],
+        models.ProcessingJob.version == job_params["version"],
     )
 
-    status_tracking = session.execute(query).first()
-    assert status_tracking.status == models.Status.SUCCEEDED
-
-    # Test for file that is not in status table
-    event["detail"]["container"]["command"][1] = "swe"
-    result = indexer.lambda_handler(event=event, context={})
-    assert result["statusCode"] == 200
-
-    query = select(models.StatusTracking.__table__).where(
-        models.StatusTracking.instrument == "swe"
-    )
-
-    status_tracking = session.execute(query).first()
-    assert status_tracking.status == models.Status.SUCCEEDED
+    processing_job = session.execute(query).first()
+    assert processing_job.id == job_id
+    assert processing_job.status == models.Status.SUCCEEDED
 
 
 def test_s3_event(session, events_client):

